staging/src/k8s.io/client-go/util/certificate/certificate_manager.go

```diff
@@ -348,10 +348,12 @@ func getCurrentCertificateOrBootstrap(
 
 	currentCert, err := store.Current()
 	if err == nil {
-		// if the current cert is expired, fall back to the bootstrap cert
-		if currentCert.Leaf != nil && time.Now().Before(currentCert.Leaf.NotAfter) {
+		if currentCert.Leaf != nil && time.Now().Before(currentCert.Leaf.NotAfter) && time.Now().After(currentCert.Leaf.NotBefore) {
```

Comment:
the local clock would have to be off by days/months for the `time.Now().Before(currentCert.Leaf.NotAfter)` comparison to fail in a problematic way, but will the new comparison make a client whose clock is a few seconds fast fail to use a just-issued cert?

Reply:
> the local clock would have to be off by days/months for the `time.Now().Before(currentCert.Leaf.NotAfter)` comparison to fail in a problematic way, but will the new comparison make a client whose clock is a few seconds fast fail to use a just-issued cert?

Yes, it will rotate

> Thanks for the change, not sure if we need to add leeway to prevent a client whose clock is a few seconds fast from refusing to use a just-issued cert and failing. Also need to exercise this condition in a test.

Thanks Reply.

This irreversible failure can easily be reproduced on the one node with master(client and server on one ). 
I use ansible to modify the system time of master and node at the same time.

When the system time is seted to exceed the validity period of the rotation certificate, restart the kubelet to make it effective, and then set it back to the current time, and it will not be able to self-heal by restarting the kubelet.

`
$ ll /etc/kubernetes/certs/
total 32
-r-------- 1 root root 1354 1月  29 16:08 ca.crt
-r-------- 1 root root 1679 1月  29 16:08 ca.key
-r-------- 1 root root 1257 1月  29 16:08 client.crt
-r-------- 1 root root 1675 1月  29 16:08 client.key
-r-------- 1 root root 1232 1月  29 16:09 kubelet-client-2021-01-29-16-09-42.pem
-r-------- 1 root root 1236 1月  30  2022 kubelet-client-2022-01-30-00-00-24.pem
lrwxrwxrwx 1 root root   86 1月  30  2022 kubelet-client-current.pem -> /etc/kubernetes/certs/kubelet-client-**2022-01-30-00-00-24.pem**
-r-------- 1 root root 1289 1月  29 16:08 server.crt
-r-------- 1 root root 1679 1月  29 16:08 server.key

one node
$ tail /paasdata/op-log/k8s/kubelet.klog
E0129 16:30:10.305421    1042 kubelet.go:2184] node "192.168.122.68" not found
E0129 16:30:10.405594    1042 kubelet.go:2184] node "192.168.122.68" not found
E0129 16:30:10.505795    1042 kubelet.go:2184] node "192.168.122.68" not found
E0129 16:30:10.605902    1042 kubelet.go:2184] node "192.168.122.68" not found
E0129 16:30:10.706051    1042 kubelet.go:2184] node "192.168.122.68" not found
E0129 16:30:10.806297    1042 kubelet.go:2184] node "192.168.122.68" not found
E0129 16:30:10.906481    1042 kubelet.go:2184] node "192.168.122.68" not found
E0129 16:30:11.006666    1042 kubelet.go:2184] node "192.168.122.68" not found
E0129 16:30:11.056598    1042 reflector.go:127] k8s.io/kubernetes/pkg/kubelet/config/apiserver.go:46: Failed to watch *v1.Pod: failed to list *v1.Pod: Unauthorized
E0129 16:30:11.106979    1042 kubelet.go:2184] node "192.168.122.68" not found

$ date +%FT%T%z
**2021-01-29T16:32:27+0800**
`

Reply:
I think there are only three cases where this can happen:
1. the issuer intentionally issued a certificate not yet valid according to their clock
2. the issuer's clock was wrong, and it issued a certificate it thought was valid `now()` but was 
  - if the apiserver clock matches the issuer clock, the certificate would actually be usable and we wouldn't want to refuse to use the certificate
  - a client detecting/refusing to use this certificate seems likely to get an identically invalid certificate if it retries and get into a loop
3. the client's clock is wrong, and it thinks the certificate is not yet valid when it actually is.

1 seems like an invalid scenario. Scenario 2 is the only one where client compensation could be correct, but between 2 and 3, 3 seems more likely to me, and the failure modes of incorrect client compensation in scenario 2 are pretty bad (avoiding using certificates that would be accepted, loop on acquiring a certificate).

Given that, I don't think this is a good change to add.

/close

Reply:
I agree with Jordan's assessment - in general it's historically been the operator's requirement to enforce clock alignment across the nodes of a cluster and Kube explicitly attempts to avoid being overly dependent on clocks.  In scenario two, the issuer clock and the node clock should be in alignment (certs are dependent on consistent clocks), and the bootstrapper is responsible for doing that.

Reply:
Combined with analysis of @liggitt @smarterclayton  , I also think that there is no need for compatibility(The situation is too extreme). Thanks again, and I wish you a good weekend. :)